### PR TITLE
[spmd] add multi-d tests for all_reduce, broadcast, scatter

### PR DESCRIPTION
**Pull request by ShawnYLuo**
add simple example tests on mult-d (3d) mesh for all_reduce, broadcast, scatter

### CODE REVIEW

File: test/spmd/tensor/test_device_mesh.py [hunk 1]
```diff
@@ -186,5 +186,136 @@ def test_device_mesh_nd(self):
...
+        local_tensor = torch.ones(3, 3) * self.rank
+        expected_ranks_by_dim = [
+            [[0, 4], [2, 6], [1, 5], [3, 7]],
+            [[0, 2], [1, 3], [4, 6], [5, 7]],
+            [[0, 1], [2, 3], [4, 5], [6, 7]],
+        ]
+        for dim, dim_ranks in enumerate(expected_ranks_by_dim):
+            for ranks in dim_ranks:
+                if self.rank in ranks:
+                    reduced_tensor = mesh.all_reduce(local_tensor, mesh_dim=dim)
```
Comment (wanchaol) [inline test/spmd/tensor/test_device_mesh.py:274]: could you explain why we need to loop through `expected_ranks_by_dim` and then call `all_reduce`?

My impression is that for n-d mesh, we should try to run all_reduce on each dim (i.e. 3d mesh, we should try to run all_reduce on mesh dim "0, 1, 2"), and then check if `local_tensor` get all_reduced on those ranks?

File: test/spmd/tensor/test_device_mesh.py [hunk 2]
```diff
@@ -186,5 +186,136 @@ def test_device_mesh_nd(self):
...
+        )
+        self.assertEqual(reduce_scattered_tensor, torch.ones(3, 3) * res_num)
+
+    @with_comms
+    def test_all_reduce_nd(self):
+        mesh_tensor = torch.arange(8).reshape(2, 2, 2)
+        mesh = DeviceMesh(self.device_type, mesh_tensor)
+        local_tensor = torch.ones(3, 3) * self.rank
+        expected_ranks_by_dim = [
+            [[0, 4], [2, 6], [1, 5], [3, 7]],
```
Comment (wanchaol) [inline test/spmd/tensor/test_device_mesh.py:267]: I am curious how do we get the this list, are we just print/visualize the 3d tensor like we did here? https://github.com/pytorch/PiPPy/blob/1f9ed8cfa93fcbb65d465912d12ad118f35fb86f/spmd/tensor/device_mesh.py#L160

File: test/spmd/tensor/test_device_mesh.py [hunk 3]
```diff
@@ -186,5 +186,136 @@ def test_device_mesh_nd(self):
...
+            [[0, 2], [1, 3], [4, 6], [5, 7]],
+            [[0, 1], [2, 3], [4, 5], [6, 7]],
+        ]
+        for dim, dim_ranks in enumerate(expected_ranks_by_dim):
```
Comment (wanchaol) [inline test/spmd/tensor/test_device_mesh.py:271]: I think ideally these ranks check should actually happen within the DeviceMesh's collective wrapper itself. We can do it as a follow up task :)

Comment (ShawnYLuo) [reply test/spmd/tensor/test_device_mesh.py:271]: "to run all_reduce on each dim (i.e. 3d mesh, we should try to run all_reduce on mesh dim "0, 1, 2")" is exactly what I'd like to do. This variable "expected_ranks_by_dim" is here to easily see the expected result so I can just sum it. Would it be more reasonable to let it be computed in mesh.get_dim_groups()?

File: test/spmd/tensor/test_device_mesh.py (hunk 2 above)
Comment (ShawnYLuo) [reply test/spmd/tensor/test_device_mesh.py:267]: Yes

> I am curious how do we get the this list, are we just print/visualize the 3d tensor like we did here?
> 
> https://github.com/pytorch/PiPPy/blob/1f9ed8cfa93fcbb65d465912d12ad118f35fb86f/spmd/tensor/device_mesh.py#L160

File: test/spmd/tensor/test_device_mesh.py (hunk 3 above)
Comment (wanchaol) [reply test/spmd/tensor/test_device_mesh.py:271]: Looking into this a bit more, I think we probably don't need to do this check logic, as when we are creating sub process groups, it's only being added when the current rank in the rank of that group, and that's why we can directly call `device_mesh.scatter(mesh_dim=1)` without checking my_rank (if my_rank not in the group, it's not calling the collective on that group so wouldn't be a issue). https://github.com/pytorch/PiPPy/blob/1608285749307c30f65dd85f4110810076e7f3cb/spmd/tensor/device_mesh.py#L175

**Comment by wanchaol** [comment]
btw https://github.com/pytorch/PiPPy/pull/322 is landed, so you can get rid of my commit and rebase to the main branch in order to get this PR merged :)
